### PR TITLE
"Upcoming docs.rs changes"

### PR DESCRIPTION
**Pull request by pietroalbini**
As we agreed in the last leadership meeting this is a blog post explaining the upcoming changes in the docs.rs build environment. The changes are implemented in https://github.com/rust-lang/docs.rs/pull/407.

We still need to schedule a day for the deployment, and we'll discuss the date in the infrastructure team meeting on 2019-09-17, hence the publishing date of the post on 2019-09-18. I want the post to be out in advance so people have time to test and adapt their crates.

As the author of the post I chose "The Rust Infrastructure Team", as I'm the one who mainly did the design and implementation work (all of this with my infra hat) and this is something infra really wants. Do y'all think it's appropriate or should I change it to another author?

r? @Mark-Simulacrum 
cc @rust-lang/core @rust-lang/infra @rust-lang/rustdoc 

Edit: oh yeah, this also includes a small commit to the blog engine to ignore vim's swapfiles.

### CODE REVIEW

File: posts/2019-09-18-upcoming-docsrs-changes.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+all the dependencies previously installed in the legacy build environment.
+
+To ensure we can continue operating the service in the future and to increase
+its reliability we also improved the sandbox builds are executed into, adding
+new limits:
+
+* Each platform will now have **15 minutes** to build its dependencies and
+  documentation.
+* **3 GB of RAM** will be available for the build.
+* Network access will be completly **disabled**.
```
Comment (Mark-Simulacrum) [inline posts/2019-09-18-upcoming-docsrs-changes.md]: Maybe worth noting here that we'll run `cargo fetch` before hand?

File: posts/2019-09-18-upcoming-docsrs-changes.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+* Only the `target/` directory will be writable, and it will be purged after
+  each build.
+
+Finally, docs.rs will now use the latest nightly available when building
+crates.
```
Comment (Mark-Simulacrum) [inline posts/2019-09-18-upcoming-docsrs-changes.md]: We should make a note to update this post after deploying the changes to link to the limits page here.

File: posts/2019-09-18-upcoming-docsrs-changes.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+```
+cargo doc --no-deps
+```
+
+If a dependency is missing please [open an issue][crates-build-env-issue] on
+the Docker image's repository.
+
+If your crate fails to build because it took more than 15 minutes to generate
+its docs or it uses more than 3 GB of RAM please [open an issue][docsrs-issue]
```
Comment (Mark-Simulacrum) [inline posts/2019-09-18-upcoming-docsrs-changes.md]: Hm, I presume that the docker command above is not limited to these specs? Maybe we can provide an approximation to it or literally the limits we're using in some relatively easy fashion?

File: posts/2019-09-18-upcoming-docsrs-changes.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+## What will change
+
+Builds will be executed inside the [rustops/crates-build-env] Docker image.
+That image contains a lot of system dependencies installed to ensure we can
+build as many crates as possible. It's already used by [Crater], and we added
+all the dependencies previously installed in the legacy build environment.
+
+To ensure we can continue operating the service in the future and to increase
+its reliability we also improved the sandbox builds are executed into, adding
```
Comment (shepmaster) [inline posts/2019-09-18-upcoming-docsrs-changes.md]:
```suggestion
its reliability we also improved the sandbox builds are executed in, adding
```

File: posts/2019-09-18-upcoming-docsrs-changes.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+new limits:
+
+* Each platform will now have **15 minutes** to build its dependencies and
+  documentation.
+* **3 GB of RAM** will be available for the build.
+* Network access will be completly **disabled**.
+* Only the `target/` directory will be writable, and it will be purged after
+  each build.
+
+Finally, docs.rs will now use the latest nightly available when building
```
Comment (shepmaster) [inline posts/2019-09-18-upcoming-docsrs-changes.md]: It is my understanding that it already uses nightly. Perhaps the change is that the version of nightly is now automatically updated each day?

File: posts/2019-09-18-upcoming-docsrs-changes.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+Once you're in a shell you can install [rustup] (it's not installed in the
+image), install Rust nightly, clone your crate's repository and then build the
+documentation:
+
+```
+cargo doc --no-deps
+```
+
+If a dependency is missing please [open an issue][crates-build-env-issue] on
```
Comment (shepmaster) [inline posts/2019-09-18-upcoming-docsrs-changes.md]:
```suggestion
If a dependency is missing, please [open an issue][crates-build-env-issue] on
```

File: posts/2019-09-18-upcoming-docsrs-changes.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+the Docker image's repository.
```
Comment (shepmaster) [inline posts/2019-09-18-upcoming-docsrs-changes.md]: Link to it again here, for ease of use.

File: posts/2019-09-18-upcoming-docsrs-changes.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+If your crate fails to build because it took more than 15 minutes to generate
+its docs or it uses more than 3 GB of RAM please [open an issue][docsrs-issue]
+and we will increase the limits for your crate (when it's reasonable). We will
```
Comment (shepmaster) [inline posts/2019-09-18-upcoming-docsrs-changes.md]: Godspeed on handling these one-off requests.

Comment (shepmaster) [inline posts/2019-09-18-upcoming-docsrs-changes.md]:
```suggestion
and we will consider reasonable increases of the limits for your crate. We will
```

File: posts/2019-09-18-upcoming-docsrs-changes.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+**not** enable network access for your crate though: you'll need to change your
+crate not to require any external resource at build time.
```
Comment (shepmaster) [inline posts/2019-09-18-upcoming-docsrs-changes.md]: Would it be be useful to show how this might work? Are there any well-known crates where we could link to a git commit where they did this?

**Comment by shepmaster** [comment]
> As the author of the post I chose "The Rust Infrastructure Team", as I'm the one who mainly did the design and implementation work (all of this with my infra hat) and this is something infra really wants. Do y'all think it's appropriate or should I change it to another author?

I think you could even say "Pietro Albini and the Rust Infrastructure Team". Self-promotion is good when deserved!

File: posts/2019-09-18-upcoming-docsrs-changes.md [hunk 10]
```diff
@@ -0,0 +1,85 @@
+---
+layout: post
+title: "Upcoming docs.rs changes"
+author: The Rust Infrastructure Team
+---
+
+On September 42nd breaking changes will be deployed to the [docs.rs] build
```
Comment (CryZe) [inline posts/2019-09-18-upcoming-docsrs-changes.md]: 42nd?

Comment (pietroalbini) [reply posts/2019-09-18-upcoming-docsrs-changes.md]: Don't you like placeholder dates?

File: posts/2019-09-18-upcoming-docsrs-changes.md (hunk 5 above)
Comment (pietroalbini) [reply posts/2019-09-18-upcoming-docsrs-changes.md]: Yeah, I'll clarify.

File: posts/2019-09-18-upcoming-docsrs-changes.md (hunk 9 above)
Comment (pietroalbini) [reply posts/2019-09-18-upcoming-docsrs-changes.md]: Not sure I have examples myself. Some FFI bindings like rusqlite bundle their C libraries if you pass a feature flag, but that's not really the same thing?

File: posts/2019-09-18-upcoming-docsrs-changes.md (hunk 3 above)
Comment (pietroalbini) [reply posts/2019-09-18-upcoming-docsrs-changes.md]: No, it's not limited, and the only thing we could provide a snippet for is setting the 3 GB memory limit. We can't disable network access otherwise people can't fetch their crate's code, and we can't set a timeout as Docker doesn't support that (it's handled by rustwide).

**Comment by pietroalbini** [comment]
Addressed most of the review comments and replied to the others.

**Comment by pietroalbini** [comment]
Updated the post to mention the deploy date of the breaking changes is going to be September 30th.

File: posts/2019-09-18-upcoming-docsrs-changes.md (hunk 3 above)
Comment (XAMPPRocky) [reply posts/2019-09-18-upcoming-docsrs-changes.md]: Could we add `time` to the command? That way people will see some kind of time estimate.

File: posts/2019-09-18-upcoming-docsrs-changes.md [hunk 11]
```diff
@@ -4,7 +4,7 @@ title: "Upcoming docs.rs changes"
 author: The Rust Infrastructure Team
 ---
 
-On September 42nd breaking changes will be deployed to the [docs.rs] build
+On September 30nd breaking changes will be deployed to the [docs.rs] build
```
Comment (shepmaster) [inline posts/2019-09-18-upcoming-docsrs-changes.md]: Traditionally, it's 30**th**. 😉

File: posts/2019-09-18-upcoming-docsrs-changes.md [hunk 12]
```diff
@@ -0,0 +1,86 @@
+---
+layout: post
+title: "Upcoming docs.rs changes"
+author: The Rust Infrastructure Team
+---
+
+On September 30nd breaking changes will be deployed to the [docs.rs] build
+environment. [docs.rs] is a free service building and hosting documentation for
+all the crates published on [crates.io]. It's [open source][docsrs-source] and
+maintained by the [Rustdoc team][rustdoc-team].
```
Comment (GuillaumeGomez) [inline posts/2019-09-18-upcoming-docsrs-changes.md]: And infra team too now. I think you deserve recognition too. :)

File: posts/2019-09-18-upcoming-docsrs-changes.md [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+## What will change
+
+Builds will be executed inside the [rustops/crates-build-env] Docker image.
+That image contains a lot of system dependencies installed to ensure we can
+build as many crates as possible. It's already used by [Crater], and we added
```
Comment (GuillaumeGomez) [inline posts/2019-09-18-upcoming-docsrs-changes.md]: It'd be nice to tell users that they can provide documentation without requiring dependencies (that's what we do in gtk-rs for example). It'd prevent to have build failures.

File: posts/2019-09-18-upcoming-docsrs-changes.md (hunk 11 above)
Comment (pietroalbini) [reply posts/2019-09-18-upcoming-docsrs-changes.md]: Did I mention I'm bad at multitasking during infra meetings?

File: posts/2019-09-18-upcoming-docsrs-changes.md (hunk 13 above)
Comment (pietroalbini) [reply posts/2019-09-18-upcoming-docsrs-changes.md]: Do you have links with any example of that?

Comment (GuillaumeGomez) [reply posts/2019-09-18-upcoming-docsrs-changes.md]: Sure!

We disable the linkage in the `build.rs` [file](https://github.com/gtk-rs/sys/blob/master/gdk-sys/build.rs) and we use the provided `docs.rs` feature in the `Cargo.toml` [file](https://github.com/gtk-rs/sys/blob/master/gdk-sys/Cargo.toml#L52) to enable/disable features too.

File: posts/2019-09-18-upcoming-docsrs-changes.md (hunk 3 above)
Comment (pietroalbini) [reply posts/2019-09-18-upcoming-docsrs-changes.md]: Added memory limits and the `time` command.

**Comment by pietroalbini** [comment]
Addressed all the requested changes.

**Comment by pietroalbini** [comment]
@rust-lang/core could I get an approval from y'all before publishing the post?

**Comment by Mark-Simulacrum** [comment]
I've asked in core on Discord for approval and I or someone else will circle back here in a few hours to approve/unapprove; personally I think we should be fine to go ahead though.

File: posts/2019-09-18-upcoming-docsrs-changes.md [hunk 14]
```diff
@@ -0,0 +1,98 @@
+---
+layout: post
+title: "Upcoming docs.rs changes"
+author: The Rust Infrastructure Team
+---
+
+On September 30th breaking changes will be deployed to the [docs.rs] build
```
Comment (nikomatsakis) [inline posts/2019-09-18-upcoming-docsrs-changes.md]: I'd rephrase this to emphasize the benefits first.

---

On September 30th, we'll be upgrading docs.rs to use a stronger sandbox when building documentation. This change is not expected to effect the vast majority of users, but it is technically a breaking change, so we want to be sure everyone is aware of it.

## what is docs.rs?

[docs.rs] is a free service building and hosting documentation ... (as before)

## What will change

...

**Comment by nikomatsakis** [comment]
oh, wait this was already posted :P
